cli/src/program_v4.rs

```diff
@@ -680,87 +559,188 @@ pub fn process_program_v4_subcommand(
     buffer_signer: Option<&dyn Signer>,
+    use_rpc: bool,
 ) -> ProcessResult {
     let blockhash = rpc_client.get_latest_blockhash()?;
-    let payer_pubkey = config.payer.pubkey();
+    let payer_pubkey = config.signers[0].pubkey();
+    let authority_pubkey = config.signers[*auth_signer_index].pubkey();
 
-    let (initial_messages, balance_needed, buffer_address) =
-        if let Some(buffer_signer) = buffer_signer {
-            let buffer_address = buffer_signer.pubkey();
-            let (create_buffer_message, required_lamports) = build_create_buffer_message(
-                rpc_client.clone(),
-                config,
-                program_address,
-                &buffer_address,
-                program_data_len,
+    // Download feature set
+    let mut feature_set = FeatureSet::default();
+    for feature_ids in FEATURE_NAMES
+        .keys()
+        .cloned()
+        .collect::<Vec<Pubkey>>()
+        .chunks(MAX_MULTIPLE_ACCOUNTS)
+    {
+        rpc_client
+            .get_multiple_accounts(feature_ids)?
+            .into_iter()
+            .zip(feature_ids)
+            .for_each(|(account, feature_id)| {
+                let activation_slot = account.and_then(status_from_account);
+
+                if let Some(CliFeatureStatus::Active(slot)) = activation_slot {
+                    feature_set.activate(feature_id, slot);
+                }
+            });
+    }
+    let program_runtime_environment =
+        solana_bpf_loader_program::syscalls::create_program_runtime_environment_v1(
+            &feature_set,
+            &ComputeBudget::default(),
+            true,
+            false,
+        )
+        .unwrap();
+
+    // Verify the program
+    let upload_range =
+        upload_range.start.unwrap_or(0)..upload_range.end.unwrap_or(program_data.len());
+    const MAX_LEN: usize =
+        (MAX_PERMITTED_DATA_LENGTH as usize).saturating_sub(LoaderV4State::program_data_offset());
+    assert!(
+        program_data.len() <= MAX_LEN,
+        "Program length {} exeeds maximum length {}",
+        program_data.len(),
+        MAX_LEN,
+    );
+    assert!(
+        upload_range.end <= program_data.len(),
+        "Range end {} exeeds program length {}",
+        upload_range.end,
+        program_data.len(),
+    );
+    let executable =
+        Executable::<InvokeContext>::from_elf(program_data, Arc::new(program_runtime_environment))
+            .map_err(|err| format!("ELF error: {err}"))?;
+    executable
+        .verify::<RequisiteVerifier>()
+        .map_err(|err| format!("ELF error: {err}"))?;
+
+    // Create and add retract message
+    let mut initial_messages = Vec::default();
+    let program_account = rpc_client
+        .get_account_with_commitment(program_address, config.commitment)?
+        .value;
+    if let Some(program_account) = program_account.as_ref() {
+        if let Some(instruction) =
+            build_retract_instruction(program_account, program_address, &authority_pubkey)?
```

Comment:
In the original flow, we do retract only if `buffer_signer` is None. So a slight deviation from that flow.
This new flow can retract a program when a redeploy is happening using buffer account. So duration of retracted program will potentially be longer compared to the old flow.

Reply:
Loader-v4 mandates that the program must be retracted to be (re)deployed: https://github.com/anza-xyz/agave/blob/2c3d43636dce6a073b21556091c89e721bec4fdf/programs/loader-v4/src/lib.rs#L241

But yes, we should probably put it back in final messages not initial messages when a source buffer is used.

Reply:
Is this addressed with the current set of commits?

Reply:
Yes: https://github.com/anza-xyz/agave/pull/3428/files#diff-d3d89642ed42e8856083506486d6ba376bad6df17e4349b03ad498d68661f88aR723-R753

Reply:
We are doing it here as well: https://github.com/anza-xyz/agave/pull/3428/files#diff-d3d89642ed42e8856083506486d6ba376bad6df17e4349b03ad498d68661f88aR641

Earlier, this was done only if `buffer_signer` was `None`. i.e. Redeploy a program using original program account. Is that condition somehow checked before calling it this line above?

Reply:
It is just constructed there, and sent later in `// Create and add deploy messages`.
Yes, the intention is to always send it if the program is currently deployed.

Reply:
Should we move it closer to where it's being used?
Also, if for some reason the `build_retract_message()` returns None, we should error out, instead of just not inserting it in the initial/final messages.

Reply:
> Should we move it closer to where it's being used?

It is constructed early because it requires the program_account which is later moved out or dropped.

> Also, if for some reason the build_retract_message() returns None, we should error out, instead of just not inserting it in the initial/final messages.

If it returns `None` it means there is no need to retract the program because it already is. This is different from loader-v3 as a program can be manually retracted and remain in that state now. In order to recover from that automatically it must handle it as an `Option`.